be/src/exec/spill/spiller.cpp

```diff
@@ -0,0 +1,230 @@
+#include "exec/spill/mem_table.h"
+#include "exec/spill/spilled_stream.h"
+#include "exec/spill/spiller_path_provider.h"
+#include "gutil/port.h"
+#include "runtime/runtime_state.h"
+#include "serde/column_array_serde.h"
+
+namespace starrocks {
+SpillProcessMetrics::SpillProcessMetrics(RuntimeProfile* profile) {
+    spill_timer = ADD_TIMER(profile, "SpillTime");
+    spill_rows = ADD_COUNTER(profile, "SpilledRows", TUnit::UNIT);
+    flush_timer = ADD_TIMER(profile, "SpillFlushTimer");
+    write_io_timer = ADD_TIMER(profile, "SpillWriteIOTimer");
+    restore_rows = ADD_COUNTER(profile, "SpillRestoreRows", TUnit::UNIT);
+    restore_timer = ADD_TIMER(profile, "SpillRestoreTimer");
+}
+// Not thread safe
+class ColumnSpillFormater : public SpillFormater {
+public:
+    ColumnSpillFormater(ChunkBuilder chunk_builder) : _chunk_builder(std::move(chunk_builder)) {}
+    Status spill_as_fmt(SpillFormatContext& context, std::unique_ptr<WritableFile>& writable,
+                        const ChunkPtr& chunk) const noexcept override;
+    StatusOr<ChunkUniquePtr> restore_from_fmt(SpillFormatContext& context,
+                                              std::unique_ptr<RawInputStreamWrapper>& readable) const override;
+    Status flush(std::unique_ptr<WritableFile>& writable) const override;
+
+private:
+    size_t _spill_size(const ChunkPtr& chunk) const;
+    ChunkBuilder _chunk_builder;
+};
+
+size_t ColumnSpillFormater::_spill_size(const ChunkPtr& chunk) const {
+    size_t serialize_sz = 0;
+    for (const auto& column : chunk->columns()) {
+        serialize_sz += serde::ColumnArraySerde::max_serialized_size(*column);
+    }
+    return serialize_sz + sizeof(serialize_sz);
+}
+
+Status ColumnSpillFormater::spill_as_fmt(SpillFormatContext& context, std::unique_ptr<WritableFile>& writable,
+                                         const ChunkPtr& chunk) const noexcept {
+    size_t serialize_sz = _spill_size(chunk);
+    context.io_buffer.resize(serialize_sz);
+    DCHECK_GT(serialize_sz, 4);
+
+    auto* buff = reinterpret_cast<uint8_t*>(context.io_buffer.data());
+    UNALIGNED_STORE64(buff, serialize_sz);
+    buff += sizeof(serialize_sz);
+
+    for (const auto& column : chunk->columns()) {
+        buff = serde::ColumnArraySerde::serialize(*column, buff);
+    }
+
+    RETURN_IF_ERROR(writable->append(context.io_buffer));
+    return Status::OK();
+}
+
+StatusOr<ChunkUniquePtr> ColumnSpillFormater::restore_from_fmt(SpillFormatContext& context,
+                                                               std::unique_ptr<RawInputStreamWrapper>& readable) const {
+    size_t serialize_sz;
+    RETURN_IF_ERROR(readable->read_fully(&serialize_sz, sizeof(serialize_sz)));
+    DCHECK_GT(serialize_sz, sizeof(serialize_sz));
+    context.io_buffer.resize(serialize_sz);
+    auto buff = reinterpret_cast<uint8_t*>(context.io_buffer.data());
+    RETURN_IF_ERROR(readable->read_fully(buff, serialize_sz - sizeof(serialize_sz)));
```

Comment:
if the serialize_sz is too big, read_fully would be a performance bottleneck and blocks too long times,
can we split the entire data into blocks, then write blocks in turn, when restore the data, we read blocks in turn.

Reply:
The smallest serialization unit is the chunk, and we will separate the task elsewhere

```
        SCOPED_TIMER(_metrics.write_io_timer);
        // flush all pending result to spilled files
        size_t num_rows_flushed = 0;
        RETURN_IF_ERROR(mem_table->flush([&](const auto& chunk) {
            num_rows_flushed += chunk->num_rows();
            RETURN_IF_ERROR(_spill_fmt->spill_as_fmt(spill_ctx, writable, chunk));
            return Status::OK();
        }));
        TRACE_SPILL_LOG << "spill flush rows:" << num_rows_flushed << ",spiller:" << this;
```
the flush possiable implements will be:
```
MemTable::flush(callback) {
    ChunkSlice chunk_slice = _sorted_buffer_chunks;
    for(chunk: chunk_slice) {
        callback(chunk);
    }
}
}
```